Build fails on ppc64le due to MUST_TAIL_CALL when tail call is not possible

mcatanzaro opened the pull request:
#### 1f021868c255213f077780a27889e5f0d80b463d
<pre>
Build fails on ppc64le due to MUST_TAIL_CALL when tail call is not possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=279985">https://bugs.webkit.org/show_bug.cgi?id=279985</a>

Reviewed by Adrian Perez de Castro.

Our use of clang::musttail attribute is not supported by Clang on
ppc64le and needs to be removed. See
<a href="https://github.com/llvm/llvm-project/pull/93267">https://github.com/llvm/llvm-project/pull/93267</a> and
<a href="https://github.com/llvm/llvm-project/issues/98859">https://github.com/llvm/llvm-project/issues/98859</a> and
<a href="https://github.com/llvm/llvm-project/issues/108014">https://github.com/llvm/llvm-project/issues/108014</a> for more info.

This change is authored by Dan Horák (thank you!).

* Source/WTF/wtf/Compiler.h:

Canonical link: <a href="https://commits.webkit.org/285261@main">https://commits.webkit.org/285261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b424d0573b49033a993e25e5f59cccf7789db41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71914 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24699 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76074 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23123 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22943 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56761 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15262 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74981 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46563 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61959 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37198 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43226 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19429 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21468 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65042 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65139 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19793 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77754 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71166 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16154 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18971 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64973 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16200 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61982 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64492 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15914 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12671 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6315 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92949 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47132 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20482 "Found 3 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/verbose-failure-dont-graph-dump-availability-already-freed.js.default, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48201 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49488 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47945 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -272,10 +272,11 @@
 #if COMPILER(CLANG)
 #if __SIZEOF_POINTER__ == 8
 #if !defined(MUST_TAIL_CALL) && defined(__cplusplus) && defined(__has_cpp_attribute)
-#if __has_cpp_attribute(clang::musttail)
+#if __has_cpp_attribute(clang::musttail) && !defined(__powerpc__)
```
It seems that my submission with the Skia and skcms changes will land, see:

- https://skia-review.googlesource.com/c/skia/+/910216
- https://skia-review.googlesource.com/c/skcms/+/910656

...which means we will only need this change in `Compiler.h` from this PR in the end.

mcatanzaro replied:
Right. What we'll do here is land only the Compiler.h change via this merge request, and pick up the rest of the changes in our main branch whenever we next update skia.

I will separately land the full set of changes (sans the patching framework) on the webkitglib/2.46 branch.

mcatanzaro replied:
2.46: https://github.com/WebKit/WebKit/commit/33bea1079e80fcbff6208a6b52c027d00c23f95d

mcatanzaro committed:
Build fails on ppc64le due to MUST_TAIL_CALL when tail call is not possible
https://bugs.webkit.org/show_bug.cgi?id=279985

Reviewed by Adrian Perez de Castro.

Our use of clang::musttail attribute is not supported by Clang on
ppc64le and needs to be removed. See
https://github.com/llvm/llvm-project/pull/93267 and
https://github.com/llvm/llvm-project/issues/98859 and
https://github.com/llvm/llvm-project/issues/108014 for more info.

This change is authored by Dan Horák (thank you!).

* Source/WTF/wtf/Compiler.h:

Canonical link: https://commits.webkit.org/285261@main
